### ISSUE TITLE
A security issue in the `ModularSquareRoot` function leads to a DOS attack

### ISSUE THREAD

**Issue by roadicing** [weidai11/cryptopp#1249]
Hi, recently I found a security issue in the [ModularSquareRoot](https://github.com/weidai11/cryptopp/blob/3e3b8af96d4d957b77588be91df7715efbe2ed69/nbtheory.cpp#L544) function of `Crypto++` library that would cause an infinite loop, since this function is being used in [ECP::DecodePoint](https://github.com/weidai11/cryptopp/blob/93208e83937a351babdfd9cfa07d7908ba15f68b/ecp.cpp#L117), an attacker could potentially craft a malformed DER public key file, and any user or server attempting to read this public key file in processes such as ECDSA may be susceptible to a DOS attack.

### Issue
The issue lies in [the second while loop](https://github.com/weidai11/cryptopp/blob/3e3b8af96d4d957b77588be91df7715efbe2ed69/nbtheory.cpp#L558) in this function, the loop starts with `n = 2` and increments `n` by one each time until we find an `n` such that `Jacobi(n, p) = -1`. However, this overlooks the case when `p` is in the form of `m^2` and `m` is an odd number. In this case, we have:

```
jacobi(n, p) = jacobi(n, m^2)
             = jacobi(n, (p_1 * p_2 * ... * p_i)^2)
             = jacobi(n, p_1)^2 * jacobi(n, p_2)^2 * ... * jacobi(n, p_i)^2
             = (jacobi(n, p_1) * jacobi(n, p_2) * ... * jacobi(n, p_i))^2
```

Since this time `jacobi(n, p)` is in the form of the square of a number, this means that the its value can only be `1` or `0` but not `-1`, therefore, no matter how `n` continues to increase, it will never satisfy `Jacobi(n, p) = -1` to break out of the loop, thus leading to an infinite loop.

Furthermore, since this function is being used in `ECP::DecodePoint`, we can construct a malformed DER public key file that includes an elliptic curve parameter `p` as:
 
```
72358384006116823815439217615866351214375729203207450702838342058601772551609
```

which is the square of:

```
268995137513890432434389773128616504853
```

and repackage it into a DER file, then any attempt to read and parse this DER file in `Crypto++` will trigger an infinite loop immediately.

In addition, since the `ModularSquareRoot` function is in the branch that handles `03` compression encoding, we need to ensure that the points in the constructed DER file are in compressed form (start with `03`).

### POC

Here is a proof-of-concept script:

```
#include <crypto++/cryptlib.h>
#include <crypto++/eccrypto.h>
#include <crypto++/files.h>

using namespace std;
using namespace CryptoPP;

int main()
{
    ECDSA<ECP, SHA256>::PublicKey publicKey;

    FileSource fs("malformed-secp256r1-pub.der", true);

    publicKey.Load(fs); // infinite loop

    return 0;
}
```

The content of the malformed DER file I have constructed can be seen here: [malformed-secp256r1-pub.der.tar.gz](https://github.com/weidai11/cryptopp/files/13409764/malformed-secp256r1-pub.der.tar.gz)

### Fix
To fix this issue, the simplest method is to check whether `p` is a prime number at the very beginning of the `ModularSquareRoot` function. If it is not a prime number, then directly reject any further calculations (after all, even if it is not rejected, the result calculated in this way will be wrong). 

However, primality test can be a relatively time-consuming operation when `p` is large, so another fix method is to find a way to narrow the search range for `n` in the while loop. According to [Wedeniwski 2001](https://mathworld.wolfram.com/ExtendedRiemannHypothesis.html), the first quadratic non-residue of a number mod `p` is always less than $`\lfloor \frac{3\cdot ln^2(p)}{2} \rfloor`$ for `p > 3` if `p` is a prime number (for `p` = 3, `n` is `2`), so we can search `n` in $`[2,\lfloor \frac{3\cdot ln^2(p)}{2} \rfloor)`$, if we fail to find such an `n` that satisfies `jacobi(n, p) = -1` after searching the entire interval, then just return `false` since in this case it is highly likely that `p` is not a prime.

It should be noted that when `p` is very large, even the second fix method can be a time-consuming operation, therefore, I recommend adding an additional constraint on the size of `p`, regardless of which fix is implemented. For example, the number of bits in `p` should be less than `4096` or other similar reasonable size.

Actually there are also some other fix solutions available, but I consider choosing one of the two methods mentioned above should be sufficient.

### Optimization

Additionally, we can add some checks at the very beginning of the `ModularSquareRoot` function to quickly rule out wrong cases, for example:

- Reject any `p` that do not satisfy the condition `p >= 3` (The the smallest odd prime)
- Directly return `false` when `a_exp_b_mod_c(a, (p - 1) / 2, p)` is not equal to `1` (Euler's criterion)

**Comment by noloader** [comment]
Thanks @roadicing,

The docs for `ModularSquareRoot` already states `p` must be prime. I think the problem is in the callers. Internally, callers like `ECP::DecodePoint` need to perform the check before calling `ModularSquareRoot`, just like external callers.

**Comment by roadicing** [comment]
Hi @noloader 

Yes, considering that users are directly vulnerable to DoS attacks when using the API as shown below to parse external public key files (including parsing certificates using API A):

```
DL_PublicKey_EC<ECP> pubKey;
FileSource fs(".malformed-pub.der", true);
pubKey.Load(fs); // infinite loop
```

I think it is necessary to introduce a check for whether `p` is a prime number, at least in [ECP::DecodePoint](https://github.com/weidai11/cryptopp/blob/93208e83937a351babdfd9cfa07d7908ba15f68b/ecp.cpp#L105C6-L105C22).

**Comment by noloader** [comment]
@roadicing ,

Take a look at [9aa07aebbdc6](https://github.com/weidai11/cryptopp/commit/9aa07aebbdc6) and see if it helps.

**Comment by roadicing** [comment]
@noloader 

Yes, I think it is feasible. However, since primality testing can also become time-consuming when `p` is extremely large, a more recommended approach is to also check the size of `p` before testing its primality. Currently, the largest `p` used in NIST-recommended ECDSA parameters is 521 bits, which can serve as a reasonable upper bound.

**Comment by noloader** [comment]
> external public key or certificate is extremely large, a more recommended approach is to also check the size of p before testing its primality in ECP::DecodePoint. Currently, the largest p used in NIST-recommended ECDSA parameters is 521 bits, which can serve as a reasonable upper bound.

I disagree with this one. All security parameters must be checked before use. If you don't validate it, then you can't use it. If you can't use it, then there's no sense in even loading it.

(And I don't buy into the crap like CVE-2023-3446).

**Comment by roadicing** [comment]
Hi @noloader 

The issue is that in the current version with primality checks added, an attacker can change their approach by constructing an ECDSA public key containing a very large prime number. Consequently, any user who uses the API in the same way as before the fix will still be vulnerable to a DoS attack. The difference is that before the fix, it was an infinite loop, whereas after the fix, it becomes a pseudo-infinite loop with the duration depending on the size of the prime number. When the prime number used by the attacker is extremely large (e.g., 20,000 bits), the time consumed by the pseudo-infinite loop is still very significant.

In the current fixes applied by other algorithm libraries, they generally adopt a combination of primality checks and limiting the size of p. In fact, there are other fix approaches (such as changing the implementation of finding the smallest quadratic non-residue in the `ModularSquareRoot ` function as mentioned in the original email, OpenSSL also uses this method). However, this might be somewhat redundant from a fix perspective. Therefore, if you do not consider the time cost caused by checking excessively large prime numbers to be a threat, then the fix so far is acceptable, and thank you again for your response and fix.

**Comment by gsantner** [comment]
Guess this issue can be closed / is resolved by that <https://github.com/weidai11/cryptopp/commit/9aa07aebbdc62461c3ac32f958d7c3ab89ff6f73> is on master? 

Thank you 😄

**Comment by roadicing** [comment]
> Guess this issue can be closed / is resolved by that [9aa07ae](https://github.com/weidai11/cryptopp/commit/9aa07aebbdc62461c3ac32f958d7c3ab89ff6f73) is on master?
> 
> Thank you 😄

Hi

Firstly, the infinite loop issue in `ModularSquareRoot` has indeed been addressed by patch [9aa07ae](https://github.com/weidai11/cryptopp/commit/9aa07aebbdc62461c3ac32f958d7c3ab89ff6f73). However, to be honest, as mentioned in my previous response, this fix introduces a new risk in the form of pseudo-infinite loops due to excessive time overhead from checking large primes. Particularly, since a prime check is already performed within the `ModularSquareRoot` function, re-introducing prime checks in other functions are somewhat redundant and further increases the time overhead.

Nevertheless, the time consumed by prime checks is finite. If you do not consider the time overhead caused by these checks, then this fix could be viable. Given that the Crypto++ project currently appears to be in a phase with relatively lower update and maintenance frequency, if you need further improvements, you may manually patch the `IsPrime` function to include necessary limits on prime size, and also thanks for your reply.

**Comment by Keisial** [comment]
Doing `CRYPTOPP_ASSERT(IsPrime(p));` on the different functions, is… interesting, since it will cause a noticeable performance hit on them, while not protecting the callers, which still need to do their own checks. However, that's just for DEBUG builds, so probably ok.

However, the one line that actually protects ECP:DecodePoint() from this DOS attack when parsing an untrusted key (`if (!IsPrime(p))` on line 125), is preceded by a `CRYPTOPP_ASSERT(IsPrime(p));`. So, if we are assuming `ECP:DecodePoint()` contract includes that it may be passed untrusted input, then on a debug build, such malicious input would still cause a DoS, due to the assert failure instead of the infinite loop.

**Comment by roadicing** [comment]
> Doing `CRYPTOPP_ASSERT(IsPrime(p));` on the different functions, is… interesting, since it will cause a noticeable performance hit on them, while not protecting the callers, which still need to do their own checks. However, that's just for DEBUG builds, so probably ok.
> 
> However, the one line that actually protects ECP:DecodePoint() from this DOS attack when parsing an untrusted key (`if (!IsPrime(p))` on line 125), is preceded by a `CRYPTOPP_ASSERT(IsPrime(p));`. So, if we are assuming `ECP:DecodePoint()` contract includes that it may be passed untrusted input, then on a debug build, such malicious input would still cause a DoS, due to the assert failure instead of the infinite loop.

Yes, the existing patch still has issues. However, given that Crypto++ seems to no longer be updated, I haven’t continued reaching out.
